### PR TITLE
[READY FOR REVIEW] Alternate Job Titles

### PR DESCRIPTION
**Pull request by ike709**
:cl: ike709
add: Jobs can now have alternate titles. You can choose them in Occupation Preferences. These currently have no gameplay impact, and are just for fluff.
/:cl:

You can now choose an alternate name for your job from a predefined list, configurable in Occupation Preferences. I'm pretty sure this one works correctly, actually.

I ported this from paradise, but couldn't find a PR for it.

Everything seems to work as intended.

Does not support alternate titles for assistants, heads of staff, or Bridge Officers.

Title-based outfits will be coming in a future PR.

LIST OF CURRENT ALT JOB TITLES:
Cargo Tech: "Supply Technician","Production Technician"
Shaft Miner: "Salvager", "Explorer"
Cook: "Chef","Sous-chef"
Janitor: "Sanitation Technician", "Maid"
Clown: "Morale Technician"
Librarian: "Journalist", "Reporter"
Lawyer: "Public Defender"
Chaplain: "Priest","Pastor","Minister"
Engineer: "Maintenance Technician","Engine Technician","Electrician","Repair Specialist", "Architect", "Propulsion Engineer"
Atmos Tech: "Firefighter", "Plumber"
Medical Doctor: "EMT","Paramedic","Surgeon"
Chemist: "Pharmacist"
Scientist: "Xenobiologist", "Researcher"
Roboticist: "Mechanic"
Detective: "Forensic Technician", "Private Investigator", "Inspector"

New Occupation Preferences window. Same size as the original one.
image

### CODE REVIEW

**Comment by ike709** [comment]
image

**Comment by ike709** [comment]
@Vivalas 

Our Maintainer, which art in #coderbus,
Hallowed be thy Merging.
Thy Review come. 
Thy will be done in GitHub, 
As it is in #coderbus.
Give us this PR our daily merge.
And forgive us our runtimes,
As we forgive them that runtime against us. 
And lead us not into bugs, 
But deliver us from conflicts. 
For thine is the spess, 
The autism, and the memes, 
For weekends only. 
Wew lad.

**Comment by Super3222** [comment]
Eh, don't you think this comes off as snowflakey? People will be running around with their custom rp job names that aren't any different from their real job.

If custom outfits come along with these jobs, I guess that makes them better. The only problem is that it's going to be confusing to find the people you need, and it just doesn't look good.

Maybe it's just me. I don't think job titles that have 0 gameplay/mechanic impact should be accessible w/o the HoP.

**Comment by ike709** [comment]
I think if anything this will help make things less confusing. Like if a scientist named himself Xenobiologist, hopefully miners will bitch at him less for a plasma cutter and instead bitch at a Researcher or Scientist.

File: code/datums/datacore.dm [hunk 1]
```diff
@@ -70,9 +70,28 @@
...
+		if(t)
+			if(t.fields["name"] == name)
+				foundrecord = t
+				break
+
+	var/list/all_jobs = get_job_datums()
+
+	for(var/datum/job/J in all_jobs)
+		var/list/alttitles = get_alternate_titles(J.title)
+		if(!J)	continue
```
Comment (Vivalas) [inline code/datums/datacore.dm]: What is this even here for?

File: code/game/objects/items/weapons/cards_ids.dm [hunk 2]
```diff
@@ -81,6 +81,7 @@
 	var/list/access = list()
 	var/registered_name = null // The name registered_name on the card
 	var/assignment = null
+
```
Comment (Vivalas) [inline code/game/objects/items/weapons/cards_ids.dm]: This random line break tho

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -650,6 +656,18 @@ var/list/preferences_datums = list()
 
 	return 1
 
+/datum/preferences/proc/GetPlayerAltTitle(datum/job/job)
+	return player_alt_titles.Find(job.title) > 0 \
```
Comment (Vivalas) [inline code/modules/client/preferences.dm]: Oh god this formatting, WHYYY

File: code/modules/jobs/job_types/cargo_service.dm [hunk 4]
```diff
@@ -69,6 +69,7 @@ Cargo Technician
 	spawn_positions = 2
 	supervisors = "the quartermaster and the executive officer"
 	selection_color = "#dcba97"
+	alt_titles = list("Supply Technician","Production Officer")
```
Comment (Vivalas) [inline code/modules/jobs/job_types/cargo_service.dm]: No thanks, "officer" designations are not for common use like this.

File: code/modules/jobs/job_types/cargo_service.dm [hunk 5]
```diff
@@ -244,6 +247,7 @@ Janitor
 	spawn_positions = 1
 	supervisors = "the executive officer"
 	selection_color = "#bbe291"
+	alt_titles = list("Sanitation Technician", "Maid")
```
Comment (Vivalas) [inline code/modules/jobs/job_types/cargo_service.dm]: lewd intensifies

File: code/modules/jobs/job_types/civilian.dm [hunk 6]
```diff
@@ -13,6 +13,7 @@ Strip out?
 	spawn_positions = 1
 	supervisors = "the executive officer"
 	selection_color = "#dddddd"
+	alt_titles = list("Morale Officer")
```
Comment (Vivalas) [inline code/modules/jobs/job_types/civilian.dm]: Hm.. this is kinda memey as an officer role, how about Morale Technician or something? Again give the clown an inch and he'll take a mile. Tell the clown he's an officer and all of a sudden unspeakable things happen.

File: code/modules/jobs/job_types/civilian.dm [hunk 7]
```diff
@@ -154,6 +156,7 @@ Strip out?
 	spawn_positions = 2
 	supervisors = "the executive officer"
 	selection_color = "#dddddd"
+	alt_titles = list("Public Defender","Magistrate","Judge")
```
Comment (Vivalas) [inline code/modules/jobs/job_types/civilian.dm]: Magistrate and Judge imply some sort of power, I like Public Defender though. Maybe we could work this in but that would require more discussion.

File: code/modules/jobs/job_types/security.dm [hunk 8]
```diff
@@ -191,6 +192,7 @@ Security Officer
 	supervisors = "the head of security, and the head of your assigned department (if applicable)"
 	selection_color = "#ffeeee"
 	minimal_player_age = 7
+	alt_titles = list("Security Guard", "Marine")
```
Comment (Vivalas) [inline code/modules/jobs/job_types/security.dm]: I dunno about these. Maybe rename security officer to marine or just remove the security guard one.

File: code/modules/jobs/jobs.dm [hunk 9]
```diff
@@ -130,7 +130,27 @@ var/list/nonhuman_positions = list(
...
+		occupations += job
+
+	return occupations
+
+/proc/get_alternate_titles(var/job)
+	var/list/jobs = get_job_datums()
+	var/list/titles = list()
+
+	for(var/datum/job/J in jobs)
+		if(!J)	continue
```
Comment (Vivalas) [inline code/modules/jobs/jobs.dm]: This again.

File: code/modules/jobs/job_types/cargo_service.dm (hunk 5 above)
Comment (ike709) [inline code/modules/jobs/job_types/cargo_service.dm]: Is this an actual complaint or a meme comment?

File: code/modules/jobs/job_types/cargo_service.dm (hunk 4 above)
Comment (ike709) [inline code/modules/jobs/job_types/cargo_service.dm]: Tell that to Floyd. He suggested it.

File: code/datums/datacore.dm (hunk 1 above)
Comment (ike709) [inline code/datums/datacore.dm]: Putting the right shit in the records I guess?

File: code/modules/client/preferences.dm (hunk 3 above)
Comment (ike709) [inline code/modules/client/preferences.dm]: Provide alternatives or don't complain, I don't know what I'm doing.

File: code/modules/jobs/job_types/civilian.dm (hunk 6 above)
Comment (ike709) [inline code/modules/jobs/job_types/civilian.dm]: FIne

File: code/modules/jobs/job_types/security.dm (hunk 8 above)
Comment (ike709) [inline code/modules/jobs/job_types/security.dm]: Blame Floyd. He suggested them.

File: code/modules/jobs/jobs.dm (hunk 9 above)
Comment (ike709) [inline code/modules/jobs/jobs.dm]: Explain damn it.

File: code/modules/client/preferences.dm (hunk 3 above)
Comment (Vivalas) [inline code/modules/client/preferences.dm]: Remove the slashes.

File: code/modules/jobs/job_types/cargo_service.dm (hunk 4 above)
Comment (Vivalas) [inline code/modules/jobs/job_types/cargo_service.dm]: Please remove it.

File: code/modules/jobs/jobs.dm (hunk 9 above)
Comment (Vivalas) [inline code/modules/jobs/jobs.dm]: This checks for something which should already be here

**Comment by ike709** [comment]
There you go.

**Pull request merged**